sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1572,63 +1679,31 @@ static void _send(gnrc_netif_t *netif, gnrc_pktsnip_t *pkt, bool push_back)
         }
     }
 
-    /* Split off the TX sync snip */
-    gnrc_pktsnip_t *tx_sync = IS_USED(MODULE_GNRC_TX_SYNC)
-                            ? gnrc_tx_sync_split(pkt) : NULL;
+#if IS_USED(MODULE_GNRC_TX_SYNC)
+    /* Don't confuse netdevs with TX sync snip */
+    gnrc_pktsnip_t *tx_sync = gnrc_tx_sync_split(pkt);
+#endif
     res = netif->ops->send(netif, pkt);
-    if (tx_sync != NULL) {
-        uint32_t err = (res < 0) ? -res : GNRC_NETERR_SUCCESS;
-        gnrc_pktbuf_release_error(tx_sync, err);
-    }
 
-    /* no frame was transmitted */
     if (res < 0) {
-        DEBUG("gnrc_netif: error sending packet %p (code: %i)\n",
-              (void *)pkt, res);
-
-        if (IS_USED(MODULE_NETSTATS_NEIGHBOR)) {
-            netstats_nb_update_tx(&netif->netif, NETSTATS_NB_BUSY, 0);
-        }
-    }
-#ifdef MODULE_NETSTATS_L2
-    else {
-        netif->stats.tx_bytes += res;
-    }
+#if IS_USED(MODULE_GNRC_TX_SYNC)
+        gnrc_pkt_append(pkt, tx_sync);
 #endif
-#if IS_USED(MODULE_GNRC_NETIF_PKTQ)
-    if (res == -EBUSY) {
-        int put_res;
-
-        /* Lower layer was busy.
-         * Since "busy" could also mean that the lower layer is currently
-         * receiving, trying to wait for the device not being busy any more
-         * could run into the risk of overriding the received packet on send
-         * Rather, queue the packet within the netif now and try to send them
-         * again after the device completed its busy state. */
-        if (push_back) {
-            put_res = gnrc_netif_pktq_push_back(netif, pkt);
-        }
-        else {
-            put_res = gnrc_netif_pktq_put(netif, pkt);
-            gnrc_netif_pktq_sched_get(netif);
-        }
-        if (put_res == 0) {
-            DEBUG("gnrc_netif: (re-)queued pkt %p\n", (void *)pkt);
-            return; /* early return to not release */
-        }
-        else {
-            LOG_ERROR("gnrc_netif: can't queue packet for sending\n");
-            /* If we got here, it means the device was busy and the pkt queue
-             * was full. The packet should be dropped here anyway */
-            gnrc_pktbuf_release_error(pkt, ENOMEM);
-        }
-        return;
+        _tx_failed(netif, res, pkt, push_back);
+        return NULL;
     }
-    else {
-        /* remove previously held packet */
-        gnrc_pktbuf_release(pkt);
+    if (!netif->dev->driver->confirm_send) {
+        /* success and legacy, blocking netdev API */
+#if IS_USED(MODULE_GNRC_TX_SYNC)
+        gnrc_pktbuf_release(tx_sync);
+#endif
+        return _tx_succeeded(netif, res);
```

Comment:
Should there be a release to drop the previously held packet?
```suggestion C
#if IS_USED(MODULE_GNRC_NETIF_PKTQ)
if (!netif->dev->driver->confirm_send) {
    /* remove previously held packet */
    gnrc_pktbuf_release(pkt);
}
#endif
return _tx_succeeded(netif, res);
```

Or the packet is passed to, and released in `_tx_succedded()`, because in `_process_events_await_msg()` the packet is released right after `tx_succeeded()` was called.

Reply:
Yes, but only when pktq is used. With the legacy API the netdev (or rather the netdev adaption layer) is in charge of releasing the packet.